### PR TITLE
Reduce rerenders in article feed

### PR DESCRIPTION
**Pull request by jonathonherbert**
## What's changed?

This PR reduces the amount of work we do rerendering the article feed when we poll for updates:

- articles are only added to the state when their modified date has changed, so they don't rerender unnecessarily
- the list only rerenders when its order changes (value comparison)
- the feed controls only rerender when pagination data changes (value comparison)

Scripting time comparison, in ms, across average of three measurements:

| Before | After |
|--------|-------|
| 678ms  | 7ms   |

### General
- [x] 🤖 Relevant tests added
- [x] ✅ CI checks / tests run locally
- [x] 🔍 Checked on CODE

### Client
- [x] 🚫 No obvious console errors on the client (i.e. React dev mode errors)
- [x] 🎛️ No regressions with existing user interactions (i.e. all existing buttons, inputs etc. work)
- [ ] 📷 Screenshots / GIFs of relevant UI changes included

### CODE REVIEW

File: client-v2/src/fixtures/initialState.ts [hunk 1]
```diff
@@ -296,10 +296,10 @@ const shared = {
         type: 'fixed/small/slow-IV'
       }
     },
-    pagination: null,
-    lastError: null,
-    error: null,
-    lastFetch: null,
+    pagination: undefined,
```
Comment (jonathonherbert) [inline client-v2/src/fixtures/initialState.ts]: We don't use null types much elsewhere in the project, so I updated them here for consistency's sake -- sorry for the noise!

File: client-v2/src/bundles/capiFeedBundle.ts [hunk 2]
```diff
@@ -5,17 +5,18 @@ import { previewCapi, liveCapi } from 'services/frontsCapi';
...
 
 const {
   actions: liveActions,
   reducer: capiLiveFeed,
   selectors: liveSelectors
 } = createAsyncResourceBundle<FeedState>('capiLiveFeed', {
-  indexById: false,
   selectLocalState: state => state.feed.capiLiveFeed,
-  initialData: []
+  indexById: true
```
Comment (jonathonherbert) [inline client-v2/src/bundles/capiFeedBundle.ts]: Instead of storing our feed articles in an array, we store them in an object and keep the incoming order elsewhere. This lets us update articles individually when we poll for changes, rather than having to update all of them at once.

File: client-v2/src/bundles/capiFeedBundle.ts [hunk 3]
```diff
@@ -58,7 +58,7 @@ const fetchResourceOrResults = async (
   return {
     results: checkIsContent(response) ? [response.content] : response.results,
     pagination: checkIsContent(response)
-      ? null
+      ? undefined
```
Comment (akash1810) [inline client-v2/src/bundles/capiFeedBundle.ts]: Best. Diff. Ever.

File: client-v2/src/components/FrontsCAPIInterface/Feed.tsx [hunk 4]
```diff
@@ -1,35 +1,81 @@
...
+import { State } from 'types/State';
+import { connect } from 'react-redux';
 
 interface ErrorDisplayProps {
-  error: string | null;
+  error?: string;
   children: React.ReactNode;
 }
 
+interface FeedContainerProps {
```
Comment (akash1810) [inline client-v2/src/components/FrontsCAPIInterface/Feed.tsx]: This type is very broad. Will we ever have both `liveArticleIds` and `previewArticleIds` and `prefillArticleIds`?

Would something like this be stricter:

```typescript
type Mode = 'prefill' | 'preview' | 'live';

interface FeedContainerProps {
  mode: Mode;
  articleIds: string[];
  error: string | undefined;
}
```

File: client-v2/src/components/util/RefreshPeriodically.tsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import React, { useEffect, useState } from 'react';
+
+const refreshPeriodically = ({
+  timeMs,
```
Comment (akash1810) [inline client-v2/src/components/util/RefreshPeriodically.tsx]: minor: would `rate` is more intention revealing here?

File: client-v2/src/shared/util/externalArticle.ts [hunk 6]
```diff
@@ -48,3 +50,22 @@ export function hasMainMediaVideoAtom(
...
   );
 }
+
+/**
+ * Create a selector to answer the question -- is the article last modified field older than the given date?
+ * This function is liberal in what it accepts -- if either date is invalid/missing, it returns `true`.
+ */
+export const createSelectIsArticleStale = <State>(
+  selectArticleById: (state: State, id: string) => ExternalArticle | undefined
+) => (state: State, id: string, dateStr: string | undefined): boolean => {
```
Comment (akash1810) [inline client-v2/src/shared/util/externalArticle.ts]: Could we type `dateStr` as a `Date`?

File: client-v2/src/components/FrontsCAPIInterface/Feed.tsx (hunk 4 above)
Comment (jonathonherbert) [reply client-v2/src/components/FrontsCAPIInterface/Feed.tsx]: Yes, this is a hangover from code previously in `FeedContainer`. I moved it without touching but it's very obviously due a refactor! I'll take a look, I suspect at the time I was keen to avoid expanding the scope of the PR.

File: client-v2/src/shared/util/externalArticle.ts (hunk 6 above)
Comment (jonathonherbert) [reply client-v2/src/shared/util/externalArticle.ts]: We can, but that would be mean every piece of consuming code having to roll their own date creation, along with the invalid case. I'm not sure that's preferable -- what do you think? An alternative might be to detail the format of the date string in a comment -- it's the same format we expect from CAPI.

File: client-v2/src/bundles/capiFeedBundle.ts (hunk 3 above)
Comment (jonathonherbert) [reply client-v2/src/bundles/capiFeedBundle.ts]: Couldn't be clearer 🤣

File: client-v2/src/components/FrontsCAPIInterface/Feed.tsx (hunk 4 above)
Comment (akash1810) [reply client-v2/src/components/FrontsCAPIInterface/Feed.tsx]: Ah hadn't acknowledged it was a just a move. Let's keep as is then to keep a small diff.

File: client-v2/src/components/util/RefreshPeriodically.tsx (hunk 5 above)
Comment (jonathonherbert) [reply client-v2/src/components/util/RefreshPeriodically.tsx]: 👍

**Comment by ajwl** [comment]
+1 for splitting this out. Sounds useful though finding it hard to follow

**Comment by jonathonherbert** [comment]
@ajwl that's fair, it touches a few different parts of the system in non-obvious ways! I've removed the null types but I'll split this into separate PRs for the change to the bundle creator state and the rest of the code, which should make it simpler.

File: client-v2/src/components/FrontsCAPIInterface/FeedItem.tsx [hunk 7]
```diff
@@ -173,22 +187,29 @@ class FeedItem extends React.Component<FeedItemProps> {
...
-            )}
-            {article.webPublicationDate && (
-              <FirstPublished>
-                {distanceInWordsStrict(
-                  Date.now(),
-                  new Date(article.webPublicationDate)
-                )}
-              </FirstPublished>
-            )}
+            <RefreshPeriodically rateMs={collectionArticlesPollInterval}>
```
Comment (jonathonherbert) [inline client-v2/src/components/FrontsCAPIInterface/FeedItem.tsx]: Because we can't rely on this feed updating itself as before, the component keeps these times fresh by itself.
